### PR TITLE
Remove page file dir from frontend. Resolves #1052 #1241

### LINKED ISSUE #1052: Set page_file_dir for internal use only
I want to set a sub directory for wiki pages, however it seems that when I use `page_file_dir` (or `--page-file-dir` for the CLI) it exposes this directory to end-users. Eg. if I set it to "wiki" then this path will be represented in the url.

I want to set a sub directory that is only internally used. From an end-user perspective the root directory should not change.

Any way of doing this?

### PR DESCRIPTION
**Pull request by dometto**
Addresses #1052 . In combination with the changes in https://github.com/gollum/gollum-lib/pull/322, I have only the following test failure here:

```
=========================================================================================================================================================================
Failure:
  /pages/Mordor/ should NOT include the page 'Bilbo Baggins'.
  <false> is not true.
test__gollum_pages_Mordor_(Frontend_with_lotr)
/Users/dawa/Code/gollum_project/gollum/test/test_app.rb:618:in `block (2 levels) in <top (required)>'
```

I think this has to do with the fact that the `/pages` route uses the (now unfunctional) `page_file_dir` logic under the hood. @bartkamphorst, since you're most familiar with the workings of `/pages`, could you advise or come up with a fix for this?

I also tried to simplify the logic in `app.rb` but refactoring and removing some helper methods.

### CODE REVIEW

File: lib/gollum/app.rb [hunk 1]
```diff
@@ -234,7 +234,7 @@ class App < Sinatra::Base
         wikip = wiki_page(params[:splat].first)
         halt 500 if wikip.nil?
         wiki   = wikip.wiki
-        page   = wiki.paged(join_page_name(wikip.name, wikip.ext), wikip.path, exact = true)
+        page   = wikip.page
```
Comment (dometto) [inline lib/gollum/app.rb]: This line until now was incredibly inefficient: `wikip` already contains the `page`, but we were doing *another* call to `wiki.paged` nevertheless! There were a couple of instances of this. Should improve page load times a bit.

**Comment by dometto** [comment]
Another Todo: we should remove the committer logic from the upload route and add a `Wiki#write_file` method to gollum-lib, which respects page_file_dir.

File: test/test_app.rb [hunk 2]
```diff
@@ -736,3 +712,67 @@ def app
...
+    committer = Gollum::Committer.new(@wiki, options)
+    committer.add_to_index('docs', 'custom', 'css', custom_content, {normalize: false})
+    committer.after_commit do |committer, sha|
+      @wiki.clear_cache
+      committer.update_working_dir('docs', 'custom', 'css')
+    end
+    committer.commit
+    puts "FILES: " + @wiki.files.inspect
+    get 'custom.css'
+    puts last_response.inspect
```
Comment (dometto) [inline test/test_app.rb]: Remove this puts and the one above

Comment (bartkamphorst) [reply test/test_app.rb]: 🤣

**Comment by dometto** [comment]
@bartkamphorst tests for this PR are now passing. I had to modify the logic for the `/pages` (Overview) route, because it was until now relying on `page_file_dir` being exposed to the frontend (it assumed that pages and files return the page_file_dir as part of their `url_path`). Instead, I am now applying a manual `.select!` in order to filter out only the pages and files under a certain path.

This is obviously not cheap: the underlying Git operation is to list *all* the pages in the wiki, and we then select the ones whose files start with a certain path. But 'note that this is how the page file dir option is currently implemented in gollum-lib](https://github.com/gollum/gollum-lib/blob/gollum-lib-5.x/lib/gollum-lib/git_access.rb#L169)! The current implementation is thus not *much* less efficient, I think.

Ideally, we'd have a cheaper way of listings only Trees that are under a certain subpath in the repo. Then we could use that to both implement the page_file_dir option in gollum-lib, and allow the listing of files under an arbitrary path.

For now, do you think this implementation is ok?

File: lib/gollum/app.rb [hunk 3]
```diff
@@ -434,12 +432,17 @@ class App < Sinatra::Base
...
-        @results     = wiki.pages
-        @results     += wiki.files
-        @results     = @results.sort_by { |p| p.name.downcase } # Sort Results alphabetically, fixes 922
+        wiki         = wiki_new
+        @results     = wiki.tree_list
+
+        if path
+          @path = Pathname.new(path).cleanpath.to_s if path
+          check_path   = wiki.page_file_dir ? ::File.join(wiki.page_file_dir, @path, '/') : "#{@path}/"
+          @results.select!  {|p| p.path.start_with?(check_path) }
```
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/gollum/app.rb]: I'm not worried about the performance hit of this solution versus the current way of doing things. We can look for improvements in a later PR.

File: lib/gollum/app.rb [hunk 4]
```diff
@@ -434,12 +432,17 @@ class App < Sinatra::Base
...
         )?      # end the optional non-capturing group
       }x do |path|
-        @path        = extract_path(path) if path
-        wiki_options = settings.wiki_options.merge({ :page_file_dir => @path })
-        wiki         = Gollum::Wiki.new(settings.gollum_path, wiki_options)
-        @results     = wiki.pages
-        @results     += wiki.files
-        @results     = @results.sort_by { |p| p.name.downcase } # Sort Results alphabetically, fixes 922
+        wiki         = wiki_new
+        @results     = wiki.tree_list
```
[the remark this answers is not in the document]
Comment (dometto) [reply lib/gollum/app.rb]: It already existed, but it now by default returns both files and pages at once (see today's PR in gollum-lib -- I'm on the app so can't easily past the link)

File: lib/gollum/app.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (dometto) [reply lib/gollum/app.rb]: Good point -- I think 'p' was for 'page', but that doesn't make sense now that result contains both pages and files already at this point.

**Comment by dometto** [comment]
Thanks for the sanity check @bartkamphorst!

**Pull request merged**